Switch from Yarn 1 to Yarn 4

peteryates opened the pull request:
It makes sense for new projects to start on the latest version of Yarn 🧶

4 has been out for some time now, confusingly it has its own 'package manager' so we install Yarn in the usual way then specify the version we want to use with `yarn set version stable`. Even more confusingly, `yarn set version latest` targets v1, which hasn't been the latest for years. There's more info on that in [Yarn's questions and answers](https://yarnpkg.com/getting-started/qa).

![image](https://github.com/DFE-Digital/rails-template/assets/128088/6e486e68-f3d9-4a85-ac80-ee7831e1dc40)

peteryates committed:
Switch from Yarn 1 to Yarn 4

peteryates commented on yarnrc.yml:
```diff
@@ -0,0 +1 @@
+nodeLinker: node-modules
```
This is here because without it Yarn uses [PNP](https://yarnpkg.com/features/pnp) but we need to do a bit of manual copying of images out of `node_modules` so Rails can pick them up.

dcyoung-dev replied:
I've seen this upgrade a few times and it also has the line `yarnPath: .yarn/releases/yarn-4.1.1.cjs` is this something we want to add to the template?

peteryates replied:
That's set by the `run "yarn set version stable"` which runs just after this happens in the template - both lines _should_ be present after a successful run.

dcyoung-dev replied:
I wondered if setting it upfront would skip the need to run `yarn set version stable` never tried it either way 👌

dcyoung-dev commented on gitignore:
```diff
@@ -6,6 +6,7 @@
 .envrc
 .idea
 .vscode
+.yarn
```
Do we want to match the ignored files as documented here https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored

peteryates replied:
Ah yes, thank you!

dcyoung-dev reviewed (commented):
There is an additional change required to the `Dockerfile` 
`RUN yarn install --frozen-lockfile --check-files` should change to `RUN yarn install --immutable` I'm not 100% sure about the `--check-files` flag?
[docs](https://arc.net/l/quote/dunutjhq)

peteryates committed:
Replace blanket .yarn ignore with recommended ones

Yarn's 'getting started' guide has a list of files it recommends
ignoring, so we'll use that instead.

https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored

Co-authored-by: dcyoung-dev <david@dcyoung.dev>

peteryates commented:
Ah yeah, removing `--check-files` rings a bell from our upgrading of ECF last year. [Apparently it's no longer needed](https://github.com/yarnpkg/berry/discussions/4427#discussioncomment-2684843).

peteryates committed:
Update yarn install step in the Dockerfile

The old arguments --frozen-lockfile and --check-files are now
deprecated, the guide suggests we use --immutable instead.

https://yarnpkg.com/migration/guide#migration-steps

Co-authored-by: dcyoung-dev <david@dcyoung.dev>

dcyoung-dev reviewed (approved):
Awesome upgrade 🎉
